cpp/src/io/parquet/writer_impl.cu

```diff
@@ -92,6 +113,128 @@ __global__ void stringdata_to_nvstrdesc(gpu::nvstrdesc_s *dst,
   }
 }
 
+__device__ size_type count_nested(column_device_view const &col, size_type idx)
+{
+  auto child_col         = col.child(lists_column_view::child_column_index);
+  auto offset_col        = col.child(lists_column_view::offsets_column_index);
+  size_type start_offset = offset_col.element<size_type>(idx + col.offset());
+  size_type end_offset   = offset_col.element<size_type>(idx + col.offset() + 1);
+
+  // Dremel encoding needs one set of repetition/definition values for null/empty records
+  if (start_offset == end_offset) { return 1; }
+  if (child_col.type().id() != type_id::LIST) { return end_offset - start_offset; }
+
+  size_type result = 0;
+  for (size_type i = start_offset; i < end_offset; i++) { result += count_nested(child_col, i); }
+  return result;
+}
+
+__global__ void dremel_encoding_size(column_device_view const col, size_type *values_in_row)
+{
+  // launched one thread per row
+  size_type row = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (row < col.size()) { values_in_row[row] = count_nested(col, row); }
+}
+
+__device__ void store_nested_rep_level(column_device_view const &col,
+                                       size_type idx,
+                                       uint32_t *rep_level,
+                                       size_type &cur_idx,
+                                       uint32_t cur_rep_level,
+                                       uint32_t depth)
+{
+  if (col.type().id() != type_id::LIST) {
+    rep_level[cur_idx++] = cur_rep_level;
+    return;
+  }
+
+  auto child_col         = col.child(lists_column_view::child_column_index);
+  auto offset_col        = col.child(lists_column_view::offsets_column_index);
+  size_type start_offset = offset_col.element<size_type>(idx + col.offset());
+  size_type end_offset   = offset_col.element<size_type>(idx + col.offset() + 1);
+
+  if (start_offset == end_offset) {
+    rep_level[cur_idx++] = cur_rep_level;
+    return;
+  }
+
+  ++depth;
+  for (size_type i = start_offset; i < end_offset; i++) {
+    store_nested_rep_level(child_col, i, rep_level, cur_idx, cur_rep_level, depth);
```

Comment:
That is correct. The rep level indicates at which level did we see a value again. For a `list<int>` with 2 rows:
```
[2,3,4]
[1,5,7,9]
```
The repetition values would be:
```
[0,1,1]
[0,1,1,1]
```
Which laid out flat would help us find row start points
```
0,1,1,0,1,1,1
^     ^
```

Reply:
Now that I look at it, the difference would be larger if nested. Thanks for the explanation.

I think the code could be made clearer: declare a new variable for the next level depth, or use `depth+1`, because the depth of the current element hasn't changed and we should not be incrementing `depth`. 
Also, be explicit about the `cur_rep_level` value here, something like
`store_nested_rep_level(child_col, i, rep_level, cur_idx, (i == start_offset) ? cur_rep_level : depth + 1, depth + 1);`